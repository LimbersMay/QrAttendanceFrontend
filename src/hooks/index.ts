export * from './useForm';
export * from './useUiStore';
export * from './useAuthStore';
export * from './useCheckAuth';
export * from './useGroupStore';
export * from './useQrCodeStore';
export * from './useRegistryStore';
export * from './useQrAttendanceStore';