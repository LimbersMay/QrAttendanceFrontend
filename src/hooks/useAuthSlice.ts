import {useAppDispatch, useAppSelector} from "../store";
import {checkingCredentials, login, logout, selectAuth} from "../store/auth";
import {qrAttendanceApi} from "../api/qrAttendanceApi";
import {onSetActiveGroup, setGroups, setQrCodes, setRegistries} from "../store/qrAttendance";

interface CreatingUserProps {
    name: string;
    email: string;
    lastname: string;
    password: string;
}

export const useAuthSlice = () => {

    const dispatch = useAppDispatch();
    const { displayName } = useAppSelector(selectAuth);

    const startCreatingUser = async ({name, email, lastname, password}: CreatingUserProps) => {
        // async code here
        dispatch(checkingCredentials());

        try {
            await qrAttendanceApi.post('/auth/register', {
                name,
                email,
                lastname,
                password
            });

            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            });

            const { body: user } = response.data;

            // sync code here
            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));
        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogin = async (email: string, password: string) => {
        dispatch(checkingCredentials());

        try {
            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            }, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const { body: user } = response.data;


            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));

        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogout = async () => {
        await qrAttendanceApi.post('/auth/logout', {}, {
            withCredentials: true,
            headers: {
                'Content-Type': 'application/json'
            }
        });

        dispatch(logout(null));
        dispatch(onSetActiveGroup(null));
        dispatch(setGroups([]));
        dispatch(setQrCodes([]));
        dispatch(setRegistries([]));
    }

    return {
        // properties
        displayName,

        // methods
        startCreatingUser,
        startLogin,
        startLogout
    }
}