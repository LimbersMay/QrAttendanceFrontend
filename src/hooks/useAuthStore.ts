import {useAppDispatch, useAppSelector} from "../store";
import {checkingCredentials, login, logout, selectAuth} from "../store/auth";
import {qrAttendanceApi} from "../api/qrAttendanceApi";
import {onSetActiveGroup, setGroups, setQrCodes, setRegistries} from "../store/qrAttendance";

export interface CreatingUserProps {
    name: string;
    email: string;
    lastname: string;
    password: string;
}

export const useAuthStore = () => {

    const dispatch = useAppDispatch();
    const { displayName, errorMessage, status } = useAppSelector(selectAuth);

    const startCreatingUser = async ({name, email, lastname, password}: CreatingUserProps) => {
        // async code here
        dispatch(checkingCredentials());

        try {

            // 1. First, create the user
            await qrAttendanceApi.post('/auth/register', {
                name,
                email,
                lastname,
                password
            });

            // 2. Then, login the user
            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            });

            const { user } = response.data;

            // sync code here
            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));
        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogin = async (email: string, password: string) => {
        dispatch(checkingCredentials());

        try {
            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            }, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const { user } = response.data;

            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));

        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogout = async () => {
        await qrAttendanceApi.post('/auth/logout', {}, {
            withCredentials: true,
            headers: {
                'Content-Type': 'application/json'
            }
        });

        dispatch(logout(null));
        dispatch(onSetActiveGroup(null));
        dispatch(setGroups([]));
        dispatch(setQrCodes([]));
        dispatch(setRegistries([]));
    }

    return {
        // properties
        displayName,
        errorMessage,
        status,

        // methods
        startCreatingUser,
        startLogin,
        startLogout
    }
}