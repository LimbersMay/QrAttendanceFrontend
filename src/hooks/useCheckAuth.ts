import {useSelector} from "react-redux";
import {login, logout, selectAuth} from "../store/auth";
import {useAppDispatch} from "../store";
import {useEffect} from "react";
import {qrAttendanceApi} from "../api/qrAttendanceApi";

export const useCheckAuth = () => {

    const { status } = useSelector(selectAuth);
    const dispatch = useAppDispatch();

    useEffect(() => {

        qrAttendanceApi.get("/auth/authenticated", {
            withCredentials: true
        }).then((response) => {

            const { user } = response.data;

            if (!user) {
                return dispatch(logout(null));
            }

            const { id, name, email } = user;
            dispatch(login({
                uid: id,
                displayName: name,
                email: email,
            }));
        }).catch(() => {
            dispatch(logout(null));
        });

    }, []);

    return status;
}
