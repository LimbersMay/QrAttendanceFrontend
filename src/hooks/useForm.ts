import React, {useEffect, useMemo, useState} from "react";

type FormValidation<T> = Record<string, string | null>

export type FormValidations = Record<string, [Function, string, string?]>

type FormState<T> = T & Record<string, any>

export const useForm = <T>( initialForm: FormState<T>, formValidations: FormValidations = {} ) => {

    const [formState, setFormState] = useState<FormState<T>>( initialForm );
    const [ formValidation, setFormValidation ] = useState<FormValidation<T>>({});

    useEffect(() => {
        createValidators();
    }, [formState]);

    useEffect(() => {
        setFormState(initialForm);
    }, [initialForm]);

    const isFormValid = useMemo(() => {

        for (const formValue of Object.keys(formValidation)) {
            if (formValidation[formValue] !== null) return false;
        }

        return true;
    }, [formValidation]);

    const convertValue = (value: any) => {
        return value.toString();
    }

    const onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = event.target;

        const newValue = convertValue(value);

        setFormState({
            ...formState,
            [name]: newValue
        });
    }

    const onDateChange = (name: string, value: unknown) => {
        setFormState({
            ...formState,
            [name]: value
        });
    }

    const onResetForm = () => {
        setFormState(initialForm);
    }

    const createValidators = () => {

        const formCheckedValues: FormValidation<T> = {};

        for (const formField of Object.keys(formValidations)) {
            const [ fn, errorMessage, dependency ] = formValidations[formField];

            //formCheckedValues[`${formField}Valid`] = fn(formState[formField]) ? null : errorMessage;
            // make validations that accept n params
            formCheckedValues[`${formField}Valid`] = fn(formState[formField], formState[dependency ?? '']) ? null : errorMessage;
        }

        setFormValidation( formCheckedValues );
    }

    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,
        onDateChange,
        ...formValidation,
        formValidation,
        isFormValid
    }
}