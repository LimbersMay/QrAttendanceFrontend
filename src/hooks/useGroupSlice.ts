import {useAppDispatch, useAppSelector} from "../store";
import {selectGroup, startUpdateGroup} from "../store/qrAttendance";
import {Group} from "../qrAttendance/interfaces";

export const useGroupSlice = () => {

    const dispatch = useAppDispatch();
    const { groups, active } = useAppSelector(selectGroup);

    const updateGroup = (group: Group) => {
        dispatch(startUpdateGroup(group));
    }

    return {
        // properties
        groups,
        active,

        // methods
        updateGroup
    }
}
