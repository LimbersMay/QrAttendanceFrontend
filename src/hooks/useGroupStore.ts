import {useAppDispatch, useAppSelector} from "../store";
import {
    selectGroup,
    onSetActiveGroup,
    updateGroup,
    addEmptyGroup,
    setGroups
} from "../store/qrAttendance";
import {Group} from "../qrAttendance/interfaces";
import {qrAttendanceApi} from "../api/qrAttendanceApi";

export const useGroupStore = () => {

    const dispatch = useAppDispatch();
    const {groups, active} = useAppSelector(selectGroup);

    const startUpdateGroup = async (group: Group) => {
        // async code here
        const {id, name} = group;

        await qrAttendanceApi.put('/group/update', {
            id: id,
            updatedFields: {
                name
            }
        });

        // sync code here
        dispatch(updateGroup(group));
    }

    const startNewGroup = async () => {

        // async code here
        const response = await qrAttendanceApi.post('/group/create', {
            name: 'Default'
        });

        const {body: group} = response.data;

        const newGroup = {
            id: group.id,
            date: group.createdAt,
            name: group.name
        }

        // sync code here
        dispatch(addEmptyGroup(newGroup));
        dispatch(onSetActiveGroup(newGroup));
    }

    const startLoadingGroups = async () => {
        // async code here
        const response = await qrAttendanceApi.get(`/group/all`);
        const {body: groups} = response.data;

        // sync code here
        const mappedGroups: Group[] = groups.map((group: any) => ({
            id: group.id,
            name: group.name,
            date: group.createdAt
        }));

        dispatch(setGroups(mappedGroups));
    }

    const setActiveGroup = (group: Group) => {
        dispatch(onSetActiveGroup(group));
    }

    return {
        // properties
        groups,
        active,

        // methods
        startUpdateGroup,
        setActiveGroup,
        startNewGroup,
        startLoadingGroups
    }
}
