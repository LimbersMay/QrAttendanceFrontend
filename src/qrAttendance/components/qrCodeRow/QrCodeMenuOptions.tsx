import React from 'react';
import {Divider, IconButton, Menu, MenuItem, Typography} from "@mui/material";
import {Delete, Download, Edit, MoreVert, Visibility} from "@mui/icons-material";
import {useQrCodeStore} from "../../../hooks/useQrCodeStore";
import {QrCode} from "../../interfaces";
import {useUiStore} from "../../../hooks/useUiStore";

const ITEM_HEIGHT = 48;

export const QrCodeMenuOptions = ({handleDownload, qrCode }: { handleDownload: any, qrCode: QrCode }) => {

    const { showQrCode, toggleQrCodeModal } = useUiStore();
    const { setActiveQrCode, startDeleteQrCode } = useQrCodeStore();

    const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
    const open = Boolean(anchorEl);

    const handleClick = (event: React.MouseEvent<HTMLElement>) => {
        setAnchorEl(event.currentTarget);
        setActiveQrCode(qrCode);
    };

    const handleClose = () => {
        setAnchorEl(null);
        setActiveQrCode(null);
    };

    return (
        <>
            <IconButton
                aria-label="more"
                aria-controls="long-menu"
                aria-haspopup="true"
                onClick={handleClick}
            >
                <MoreVert />
            </IconButton>
            <Menu
                id="long-menu"
                anchorEl={anchorEl}
                keepMounted
                open={open}
                onClose={handleClose}
                PaperProps={{
                    style: {
                        maxHeight: ITEM_HEIGHT * 4.5,
                        width: '20ch',
                    },
                }}
            >
                <MenuItem onClick={showQrCode}>
                    <Visibility />
                    <Typography sx={{ml: '7px'}}>Show</Typography>
                </MenuItem>

                <MenuItem onClick={toggleQrCodeModal}>
                    <Edit />
                    <Typography sx={{ml: '7px'}}>Edit</Typography>
                </MenuItem>

                <MenuItem onClick={handleDownload}>
                    <Download />
                    <Typography sx={{ml: '7px'}}>Download</Typography>
                </MenuItem>

                <Divider />

                <MenuItem onClick={startDeleteQrCode}>
                    <Delete />
                    <Typography sx={{ml: '7px'}}>Delete</Typography>
                </MenuItem>

            </Menu>
        </>
    );
}
