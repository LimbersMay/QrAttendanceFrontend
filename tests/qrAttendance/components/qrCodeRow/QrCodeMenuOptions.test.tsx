import {fireEvent, render, screen} from "@testing-library/react";
import {QrCodeMenuOptions} from "../../../../src/qrAttendance/components/qrCodeRow/QrCodeMenuOptions";
import {qrCodes} from "../../../fixtures/qrAttendanceStates";
import {useGroupStore, useQrAttendanceStore, useQrCodeStore, useUiStore} from "../../../../src/hooks";


const mockSetActiveQrCode = jest.fn();
const mockShowQrCode = jest.fn();
const mockToggleQrCodeModal = jest.fn();
const mockSetActiveGroup = jest.fn();
const mockStartDeleteQrCodeWithDependencies = jest.fn();

jest.mock('../../../../src/hooks/useGroupStore');
(useGroupStore as jest.Mock).mockImplementation(() => ({
    setActiveGroup: mockSetActiveGroup
}));

jest.mock('../../../../src/hooks/useQrCodeStore');
(useQrCodeStore as jest.Mock).mockImplementation(() => ({
    setActiveQrCode: mockSetActiveQrCode
}));


jest.mock('../../../../src/hooks/useUiStore');
(useUiStore as jest.Mock).mockImplementation(() => ({
    showQrCode: mockShowQrCode,
    toggleQrCodeModal: mockToggleQrCodeModal
}));

jest.mock('../../../../src/hooks/useQrAttendanceStore');
(useQrAttendanceStore as jest.Mock).mockImplementation(() => ({
    startDeleteQrCodeWithDependencies: mockStartDeleteQrCodeWithDependencies
}));

describe('tests for <QrCodeMenuOptions />', () => {

    beforeEach(() => jest.clearAllMocks());

    test('should match the snapshot', () => {

        const {container} = render(
            <QrCodeMenuOptions handleDownload={() => {
            }} qrCode={qrCodes[0]}/>
        );

        expect(container).toMatchSnapshot();
    });

    test("should set the activeQrCode when the menu is clicked", () => {

        render(
            <QrCodeMenuOptions
                handleDownload={() => {}}
                qrCode={qrCodes[0]}
            />
        );

        const menuButton = screen.getByTestId('qr-code-menu-button');
        fireEvent.click(menuButton);

        expect(mockSetActiveQrCode).toHaveBeenCalledWith(qrCodes[0]);
    });

    test("should call the showQrCode function when the show button is clicked", () => {

        render(
            <QrCodeMenuOptions handleDownload={() => {
            }} qrCode={qrCodes[0]}/>
        );

        // get by aria-label
        const showButton = screen.getByLabelText('showQrCodeButton');
        fireEvent.click(showButton);

        expect(mockShowQrCode).toHaveBeenCalled();
    });

    test("should call the toggleQrCodeModal function when the edit button is clicked", () => {

            render(
                <QrCodeMenuOptions handleDownload={() => {
                }} qrCode={qrCodes[0]}/>
            );

            // get by aria-label
            const editButton = screen.getByLabelText('editQrCodeButton');
            fireEvent.click(editButton);

            expect(mockToggleQrCodeModal).toHaveBeenCalled();
    });

    test('should call the handleDownload function when the download button is clicked', () => {

        const mockHandleDownload = jest.fn();

        render(
            <QrCodeMenuOptions handleDownload={mockHandleDownload} qrCode={qrCodes[0]}/>
        );

        // get by aria-label
        const downloadButton = screen.getByLabelText('downloadQrCodeButton');
        fireEvent.click(downloadButton);

        expect(mockHandleDownload).toHaveBeenCalled();

    });

    test("should call the startDeleteQrCodeWithDependencies function when the delete button is clicked", () => {

        render(
            <QrCodeMenuOptions handleDownload={() => {
            }} qrCode={qrCodes[0]}/>
        );

        // get by aria-label
        const deleteButton = screen.getByLabelText('deleteQrCodeButton');
        fireEvent.click(deleteButton);

        expect(mockStartDeleteQrCodeWithDependencies).toHaveBeenCalled();
    });
});
