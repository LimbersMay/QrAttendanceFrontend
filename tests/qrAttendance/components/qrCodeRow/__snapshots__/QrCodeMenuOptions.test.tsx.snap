// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <QrCodeMenuOptions /> should match the snapshot 1`] = `
<div>
  <button
    aria-controls="long-menu"
    aria-haspopup="true"
    aria-label="more"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    data-testid="qr-code-menu-button"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="MoreVertIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
